zerver/lib/email_notifications.py

```diff
@@ -284,6 +286,36 @@ def get_narrow_url(user_profile: UserProfile, message: Message,
             stream = Stream.objects.only('id', 'name').get(id=message.recipient.type_id)
         return topic_narrow_url(user_profile.realm, stream, message.topic_name())
 
+def get_mentioned_group_name(messages: List[Dict[str, Any]],
+                             user_profile: UserProfile) -> Optional[str]:
+    """
+     Returns the user group name to display in the email notification
+     if user group(s) are mentioned.
+    """
+    min_user_group_size = float('inf')
+    smallest_user_group = None
+    for message in messages:
+        mentioned_group_id = message['mentioned_group']
+        if mentioned_group_id is None and message['trigger'] == 'mentioned':
+            # If a user's mentioned multiple ways, we use
+            # the highest priority i.e. personal over user group.
+            smallest_user_group = None
+            break
+        if mentioned_group_id is not None:
+            mentioned_user_group = access_user_group_by_id(mentioned_group_id, user_profile)
+            mentioned_group_size = len(get_user_group_members(mentioned_user_group))
```

Comment:
So I'm not 100% sure if this is required, but I think that we should be doing a bit of memoization here. The methods `mentioned_user_group` and `mentioned_group_size` both involve a DB query each and we are using these methods in a loop here. So as the number of messages passed to `get_mentioned_group_name` are increased, the number of DB queries we are doing might be increased unnecessarily.

A concrete example of when memoization could be useful here is if in there are some 2 or more messages performing the same group mention. E.g.

> Hey @team1 and @team2 this looks pretty important can you deal with it ASAP?

> On second though, @team2 deal with this other issue instead and let @team1 handle this.

Such a style of sending messages is definitely possible even we don't really do this in CZO.

Memoizatation can be easily achieved within this function by maintaining a dictionary mapping tuples of `(mentioned_group_id, user_profile)` to the results of `len(get_user_group_members(...))`. If the value exists use it otherwise perform these function calls and update the dictionary accordingly. The dictionary could be called `mentioned_group_size_cache` or something.

Once again, this is just an idea, I'm not 100% sure if we'd really want to do this though. @timabbott what do you think?

Reply:
Yes, agreed, this sounds very expensive as written.  I had been imagining we'd do the determinations once inside `bugdown`, not once per outgoing message.

Reply:
> Yes, agreed, this sounds very expensive as written. I had been imagining we'd do the determinations once inside `bugdown`, not once per outgoing message.

@timabbott could you please elaborate on this a bit? @hashirsarwar and I are kind of confused about what you're suggesting here.

Reply:
@hashirsarwar you also need to add a test for this block. The test you recently added does a good job of covering mentions for a single message. But we still need a test to look at how we handle detecting the right mention in multiple messages for an email notification.

Reply:
@Hypro999 I have added `_extra_context_in_missed_stream_user_group_mention` to test sent emails already. Do you want me to add a test for `get_mentioned_group_name` specially?

Reply:
The test `_extra_context_in_missed_stream_user_group_mention` doesn't seem to be as comprehensive as the test you added for the second commit where you checked multiple cases (pm mention, direct mention and group priority).

Here when sending an email, you're basically using the same kind of strategy to determine the cause of the mention, just now you're doing it over multiple messages instead of for a single message. However, this test does not cover the same range of cases as that other test (pm mention, direct mention and group priority). I think that making three methods/functions, one for each of the three cases might be the right thing to do.

So for example, one of the functions could send a message with a group mention and then a message with a personal message and then in the email it should say "you were mentioned" and not "@team was mentioned".